lib/git_worktree.rb

```diff
@@ -0,0 +1,336 @@
+    @remote_name   = 'origin'
+    @cred          = Rugged::Credentials::UserPassword.new(:username => @name,
+                                                           :password => @password)
+    @base_name     = File.basename(@path)
+    process_repo(options)
+  end
+
+  def delete_repo
+    return false unless @repo
+    @repo.close
+    FileUtils.rm_rf(@path)
+    true
+  end
+
+  def add(path, data, commit_sha = nil, default_entry_keys = {})
+    entry = {}
+    entry[:path] = path
+    ENTRY_KEYS.each { |key| entry[key] = default_entry_keys[key] if default_entry_keys.key?(key) }
+    entry[:oid]  = @repo.write(data, :blob)
+    entry[:mode] ||= DEFAULT_FILE_MODE
+    entry[:mtime] ||= Time.now
+    current_index(commit_sha).add(entry)
+  end
+
+  def remove(path, commit_sha = nil)
+    current_index(commit_sha).remove(path)
+  end
+
+  def remove_dir(path, commit_sha = nil)
+    current_index(commit_sha).remove_dir(path)
+  end
+
+  def file_exists?(path, commit_sha = nil)
+    find_entry(path, commit_sha) ? true : false
+  end
+
+  def directory_exists?(path, commit_sha = nil)
+    entry = find_entry(path, commit_sha)
+    entry && entry[:type] == :tree
+  end
+
+  def read_file(path, commit_sha = nil)
+    read_entry(fetch_entry(path, commit_sha))
+  end
+
+  def read_entry(entry)
+    @repo.lookup(entry[:oid]).content
+  end
+
+  def directory?(path, commit_sha = nil)
+    entry = fetch_entry(path, commit_sha)
+    entry[:type] == :tree
+  end
+
+  def entries(path, commit_sha = nil)
+    tree = get_tree(path, commit_sha)
+    tree.find_all.collect { |e| e[:name] }
+  end
+
+  def nodes(path, commit_sha = nil)
+    tree = path.empty? ? lookup_commit_tree(commit_sha || @commit_sha) : get_tree(path, commit_sha)
+    entries = tree.find_all
+    entries.each do |entry|
+      entry[:full_name] = File.join(@base_name, path, entry[:name])
+      entry[:rel_path] = File.join(path, entry[:name])
+    end
+  end
+
+  def save_changes(message, owner = :local)
+    cid = commit(message)
+    if owner == :local
+      lock { merge(cid) }
+    else
+      merge_and_push(cid)
+    end
+    true
+  end
+
+  def commit(message)
+    tree = @current_index.write_tree(@repo)
+    parents = @repo.empty? ? [] : [@repo.ref(HEAD_REF).target].compact
+    create_commit(message, tree, parents)
+  end
+
+  def file_attributes(fname)
+    walker = Rugged::Walker.new(@repo)
+    walker.sorting(Rugged::SORT_DATE)
+    walker.push(@repo.ref(HEAD_REF).target)
+    commit = walker.find { |c| c.diff(:paths => [fname]).size > 0 }
+    return {} unless commit
+    {:updated_on => commit.time.gmtime, :updated_by => commit.author[:name]}
+  end
+
+  def merge(commit, rebase = false)
+    master_branch = @repo.ref(HEAD_REF)
+    merge_index = master_branch ? @repo.merge_commits(master_branch.target, commit) : nil
+    if merge_index && merge_index.conflicts?
+      result = differences_with_master(commit)
+      raise MiqException::MiqGitConflicts, result
```

Comment:
The Miq prefix on GitConflicts should be removed.  In fact, the namespace of the exception should probably be GitWorktree::GitConflicts.